snowfakery/data_generator_runtime.py

```diff
@@ -119,6 +121,45 @@ class Dependency(NamedTuple):
 )
 
 
+class SlotState(Enum):
+    UNUSED = auto()
+    ALLOCATED = auto()
+    CONSUMED = auto()
+
+
+class NicknameSlot:
+    _tablename: str
+    id_manager: IdManager
+    allocated_id: int = None
+
+    def __init__(self, tablename, id_manager):
+        self._tablename = tablename
+        self.id_manager = id_manager
+
+    @property
+    def id(self):
+        if self.allocated_id is None:
+            self.allocated_id = self.id_manager.generate_id(self._tablename)
+        return self.allocated_id
```

Comment:
I'm looking forward to when we can use https://docs.python.org/3/library/functools.html#functools.cached_property for this sort of memoization (it was added in Python 3.8)